### PR TITLE
DO NOT MERGE! Serialization prototype

### PR DESCRIPTION
**Pull request by thallgren**
This is work in progress. PR is for review purposes only.

### CODE REVIEW

File: lib/puppet/pops/parser/locator.rb [hunk 1]
```diff
@@ -262,6 +266,28 @@ def extract_text(offset, length)
...
+      attributes => {
+        string => String[1],
+        file => String,
+        line_index => {
+          type => Optional[Array[Integer]],
+          value => undef
+        }
+      }
+    }
+      OBJECT
```
Comment (hlindberg) [inline lib/puppet/pops/parser/locator.rb]: don't get this `OBJECT` can you elaborate?

File: Gemfile [hunk 2]
```diff
@@ -27,6 +27,7 @@ gem "puppet", :path => File.dirname(__FILE__), :require => false
 gem "facter", *location_for(ENV['FACTER_LOCATION'] || ['> 2.0', '< 4'])
 gem "hiera", *location_for(ENV['HIERA_LOCATION'] || ['>= 2.0', '< 4'])
 gem "rake", "10.1.1", :require => false
+gem "msgpack", "~> 0.7.6", :require => false
```
Comment (joshcooper) [inline Gemfile]: We'll need to resolve the issues we had with semantic_puppet before adding a hard dependency on msgpack, e.g. puppet-agent, puppetserver, Gemfile, ext/project_data.yaml, etc.

File: lib/puppet/pops/parser/locator.rb [hunk 3]
```diff
@@ -262,6 +266,28 @@ def extract_text(offset, length)
...
+      @type = Pcore.add_object('Pcore::AST::Locator', locator_def, loader).resolve(Types::TypeParser.singleton, loader)
+      ir.register_implementation('Pcore::AST::Locator', self, loader)
+    end
```
Comment (hlindberg) [inline lib/puppet/pops/parser/locator.rb]: Why is AST in the {{Pcore}} name space as opposed to say {{Puppet}} ? While it is expressed in Pcore, it isn't Pcore itself is it?

File: lib/puppet/pops/types/p_runtime_type.rb [hunk 4]
```diff
@@ -3,8 +3,22 @@ module Types
 
 # @api public
 class PRuntimeType < PAnyType
+
   TYPE_NAME_OR_PATTERN = PVariantType.new([PStringType::NON_EMPTY, PTupleType.new([PRegexpType::DEFAULT, PStringType::NON_EMPTY])])
 
+  def self.register_ptype(loader, ir)
+    create_ptype(loader, ir, 'ScalarType',
```
Comment (hlindberg) [inline lib/puppet/pops/types/p_runtime_type.rb]: ? is a runtime Scalar? (Don't get this)

**Comment by hlindberg** [comment]
In general this looks really good. Some things are hard to follow due to the nature of how things compose into a framework to support serialization.
At this point I just had a couple of questions regarding namespacing and some things I did not understand.

File: lib/puppet/pops/parser/locator.rb (hunk 1 above)
Comment (thallgren) [inline lib/puppet/pops/parser/locator.rb]: no, but I can remove it ;-)
